openprescribing/api/views_spending.py

```diff
@@ -483,82 +503,118 @@ def tariff(request, format=None):
         err += 'date=2015-04-01'
         return Response(err, status=400)
 
-    params = [codes]
-
-    if spending_type in [BnfHierarchy.product, BnfHierarchy.presentation]:
-        assert len(codes) > 0
-        query = _get_presentations_by_practice(codes, org_ids, date)
-        params.append(org_ids)
-
-    elif spending_type in [BnfHierarchy.section, BnfHierarchy.chemical]:
-        assert len(codes) > 0
-        practice_ids = utils.get_practice_ids_from_org(org_ids)
-        query = _get_chemicals_or_sections_by_practice(codes,
-                                                       practice_ids,
-                                                       spending_type,
-                                                       date)
-        params.append(practice_ids)
-
-    else:
-        assert spending_type is None
-        assert len(codes) == 0
+    practice_ids = utils.get_practice_ids_from_org(org_ids)
+    practices = Practice.objects.order_by('code')
+    if practice_ids:
+        practices = practices.filter(code__in=practice_ids)
+    data = _get_prescribing_entries(codes, practices, 'practice', date=date)
+    return Response(list(data))
 
-        practice_ids = utils.get_practice_ids_from_org(org_ids)
-        query = _get_total_spending_by_practice(practice_ids, date)
-        params.append(practice_ids)
 
+def _get_prescribing_entries(bnf_code_prefixes, orgs, org_type, date=None):
+    db = get_db()
+    items_matrix, quantity_matrix, actual_cost_matrix = _get_prescribing_for_codes(
+        db, bnf_code_prefixes
+    )
+    # If no data at all was found, return early which results in an empty
+    # iterator
+    if items_matrix is None:
+        return
+    # Group together practice level data to the appropriate organisation level
+    grouper = group_by(org_type)
+    items_matrix = grouper(items_matrix)
+    quantity_matrix = grouper(quantity_matrix)
+    actual_cost_matrix = grouper(actual_cost_matrix)
+    # `grouper.offsets` maps each organisation's primary key to its row offset
+    # within the matrices. We pair each organisation with its row offset,
+    # ignoring those organisations which aren't in the mapping (which implies
+    # that they did not prescribe in this period)
+    org_offsets = [
+        (org, grouper.offsets[org.pk])
+        for org in orgs
+        if org.pk in grouper.offsets
+    ]
+    # Pair each date with its column offset (either all available dates or just
+    # the specified one)
     if date:
-        params.append([date])
-
-    data = utils.execute_query(query, params)
-    return Response(data)
+        date_offsets = [(date, db.date_offsets[date])]
+    else:
+        date_offsets = sorted(db.date_offsets.items())
+    # Yield entries for each organisation on each date
+    for date, col_offset in date_offsets:
+        for org, row_offset in org_offsets:
+            index = (row_offset, col_offset)
+            items = items_matrix[index]
```

Comment:
Yes, I think we definitely will want something that handles this for us. But I deliberately didn't write anything now as I want the dust to settle on this before we start baking in one particular abstraction. I think the time to do this will be when we start changing the API (or creating a new API) to support new features.